Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp

```diff
@@ -0,0 +1,235 @@
+ */
+KafkaHistoStreamDecoder::KafkaHistoStreamDecoder(
+    std::shared_ptr<IKafkaBroker> broker, const std::string &histoTopic,
+    const std::string &instrumentName)
+    : m_broker(broker), m_histoTopic(histoTopic),
+      m_instrumentName(instrumentName), m_histoStream(), m_workspace(),
+      m_buffer(), m_thread(), m_interrupt(false), m_capturing(false),
+      m_exception() {
+  // Initialize buffer workspace
+  m_workspace = createBufferWorkspace();
+}
+
+/**
+ * Destructor.
+ * Stops capturing from the stream
+ */
+KafkaHistoStreamDecoder::~KafkaHistoStreamDecoder() { stopCapture(); }
+
+/**
+ * Start capturing from the stream on a separate thread. This is a non-blocking
+ * call and will return after the thread has started
+ */
+void KafkaHistoStreamDecoder::startCapture(bool) {
+  g_log.debug() << "Starting capture on topic: " << m_histoTopic << "\n";
+  m_histoStream =
+      m_broker->subscribe({m_histoTopic}, SubscribeAtOption::LATEST);
+
+  m_thread = std::thread([this]() { this->captureImpl(); });
+  m_thread.detach();
+}
+
+/**
+ * Stop capturing from the stream. This is a blocking call until the capturing
+ * function has completed
+ */
+void KafkaHistoStreamDecoder::stopCapture() {
+  g_log.debug() << "Stopping capture\n";
+
+  // This will interrupt the "event" loop
+  m_interrupt = true;
+  // Wait until the function has completed. The background thread will exit
+  // automatically
+  while (m_capturing) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  };
+}
+
+/**
+ * Check if there is data available to extract
+ * @return True if data has been accumulated so that extractData()
+ * can be called, false otherwise
+ */
+bool KafkaHistoStreamDecoder::hasData() const {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+  return !m_buffer.empty();
+}
+
+/**
+ * Check for an exception thrown by the background thread and rethrow
+ * it if necessary.
+ * @return A pointer to the data collected since the last call to this
+ * method
+ */
+API::Workspace_sptr KafkaHistoStreamDecoder::extractData() {
+  if (m_exception) {
+    throw *m_exception;
+  }
+
+  auto workspace_ptr = extractDataImpl();
+  return workspace_ptr;
+}
+
+// -----------------------------------------------------------------------------
+// Private members
+// -----------------------------------------------------------------------------
+
+API::Workspace_sptr KafkaHistoStreamDecoder::extractDataImpl() {
+  std::lock_guard<std::mutex> lock(m_buffer_mutex);
+
+  if (!m_capturing) {
+    throw Exception::NotYet("Local buffers not initialized.");
+  }
+
+  if (m_buffer.empty()) {
+    throw Exception::NotYet("No message to process yet.");
+  }
+
+  auto histoMsg = GetEventHistogram(m_buffer.c_str());
+
+  auto shape = histoMsg->current_shape();
+  auto nbins = shape->Get(0) - 1;
+  auto nspectra = static_cast<size_t>(shape->Get(1));
+
+  // TODO: This is a hack, wrong spectra number provided
+  nspectra = m_workspace->getNumberHistograms();
```

Comment:
This should probably be removed prior to merging.

It was necessary to add it because the number of spectra in the schema is about double of what it should be, due to SANS2D (and LOKI) using a trick to enable live histogramming. @matthew-d-jones will fix the SANS2D map file, so that the correct number of spectra will be provided.

Once that is done, this line can safely be removed.

Reply:
@matthew-d-jones any progress on this?

Reply:
The latest NeXus-Streamer docker image has the fixed map file: `screamingudder/nexus-streamer:4282e9a`.